README.md

# Rppsc

<!-- badges: start -->
<!-- badges: end -->
## Description

Rppsc is a package to plot protein sequence compostion. The goal of Rppsc is to visualize the protein sequence composition regarding different composition descriptor. The 5 attributes include hydrophobiciy, van der Waals volume, polarity, polarizability, and desolvation.

## Installation

You can install the released version of Rppsc package using:

``` r
require("devtools")
install_github("dxjasmine/Rppsc", build_vignettes = TRUE)
library("Rppsc")

#To run shinyApp
runRppsc()
```
## Overview
An overview of the package is demonstrated as below.
![](./inst/extdata/A1.png)
Reference:
Xiao et al. (2015) Comprehensive toolkit for generating various numerical features of protein sequences  <doi:10.1093/bioinformatics/btv042>

## Contribution

Functions in this package includes:

- plotCG()
- plotAPCG()

The function plotCG and plotAPCG were authored by Jiayan Wang. Both functions take advantage of the functions in the protr package and make visualization extention.

plotCG() function uses sequence composition analysis function in protr package. ggplot2 package is used to generate visualization of proteins composition. For each plot, one compositon descriptor is analyzed within multiple proteins. Part of the code for plotting the circular chart are adated from the R graph gallery for formating.

plotAPCG() calculates the average of the compositon descriptor data obtained from the function from protr package. gplot package is used to generate compositon mapping for multiple proteins with multiple attributes.

##compostion mapping regarding 5 attributes
##hydrophobicity, van der waal, polarity, polarizability
##desolvation, desolvation
Here is an example which shows you how to solve a common problem:

user can use their own csv file providing pdbid and seuquence
Default data --"proSeq" are also provided

``` r
library(Rppsc)
#  example code
#default data
hydrophobicity_plot <- plotCG(file = "proSeq", type= 1, circular_plot = TRUE)
hydrophobicity_plot #call to view the plot
polarity_plot <- plotCG(file = "proSeq", type= 3, circular_plot = TRUE)
polarity_plot
# full heat map of sequence composition of 5 attributes
all_prot_comp <- plotAPCG(file = "proSeq")
all_prot_comp
```

## Reference

Xiao, N., Cao, D.-S., Zhu, M.-F., & Xu, Q.-S. (2015). protr/ProtrWeb: R package and web server for
generating various numerical representation schemes of protein sequences. Bioinformatics, 31(11), 1857–1859.
https://doi.org/10.1093/bioinformatics/btv042

The R Graph Gallery -Circular barplot with groups. (2018).
Retrieved December 1, 2019, from https://www.r-graph-gallery.com/297-circular-barplot-with-groups.html


NEWS.md
# Rppsc 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(Rppsc)

test_check("Rppsc")


tests/testthat/test-plotCG.R
context("Ploting Protein Sequence Composition")
library(Rppsc)

test_that("plotting a composiiton graph", {

  file_path = "proSeq"
  type = 1

  p <- plotCG(file = file_path, type = type)

  expect_true(is.ggplot(p))
  expect_that( p$labels$fill, equals("category"))
})

test_that("correct number of categories", {

  file_path = "proSeq"
  type = 2

  p <- plotCG(file = file_path, type = type)

  expect_that(length(unique(p$data$category)), equals(3))
})




tests/testthat/test-plotAPCG.R
context("Ploting all protein compostion")
library(Rppsc)



test_that("return a correct list object", {

  file_path = "proSeq"


  plot <-  plotAPCG(file_path)
  expect_true(is.list(plot))
})


test_that("The plot has five categories", {


  file_path = "proSeq"

  plot <-  plotAPCG(file_path)
  expect_that(length(plot$colInd), equals(5))

})


R/plotCG.R
#'Plot A protein Composition Graph
#'
#'A function that plots the composition of protein sequence given one attribute
#'at a time. Attributes include hydrophobicity, van der Waals volume, polarity,
#'polarizability, and desolvation
#'
#' @param file input csv file containing the protein name and sequence
#' @param type integer, indicates one attribute for plotting the composition
#' different Attributes are listed below:
#' 1: hydrophobicity
#' 2: van der Waals volume
#' 3: polarity
#' 4: polarizability
#' 5: desolvation
#'
#' @return
#' This function returns a composition graph given the attribute type.
#'
#' @examples
#' # defaul value is to show hydrophobicity composition in a circular plot
#' p <- plotCG(file = "proSeq",type= 1, circular_plot = TRUE )
#' p #call p to see the plot
#'
#' @references
#'The R Graph Gallery -Circular barplot with groups. (2018).Retrieved
#'December 1, 2019, from https://www.r-graph-gallery.com/297-circular-barplot-with-groups.html
#'
#' @export
#' @import ggplot2
#' @import protr
#' @import utils
plotCG <- function(file = "proSeq",type= 1, circular_plot = TRUE ) {

  library(utils)
  library(ggplot2)
  library(protr)

  #Check type value
  if(!(type<6 & type>0 & type %% 1== 0)) {
    stop("Invalid value. Should be integer between 1-5")
  }
  #validate file path
  if(!(file_test("-f",file)) & (file != "proSeq")){
      stop("Invalid file path.")
  }

  #load the default data or from validated file path
  if (file == "proSeq"){
    protein_sequence <- data.frame(lapply(proSeq, as.character), stringsAsFactors=FALSE)
  }else{
    raw_sequence <- read.csv(filepath,header = TRUE,sep = ",")
    protein_sequence <- data.frame(lapply(raw_sequence, as.character), stringsAsFactors=FALSE)
  }

  #get number of proteins in list
  protein_num <- nrow(protein_sequence)

  #all protein name
  protein_name <- NULL



  #loop over all proteins to get their protein composition by extractCTDC function
  #Data are stored in 5 categories: hydrophobiciy, van der Waals volume, polarity
  #polarizability,desolvation.
  for(i in seq_along(protein_sequence[,1]) ) {

    #initiation
    if(is.null(protein_name)){
      protein_name  <- protein_sequence[i,1]
      hydro  <- data.matrix(extractCTDC (protein_sequence[i,2]))[1:3,1]
      vdm  <- data.matrix(extractCTDC (protein_sequence[i,2]))[4:6,1]
      pol  <- data.matrix(extractCTDC (protein_sequence[i,2]))[7:9,1]
      polabil  <- data.matrix(extractCTDC (protein_sequence[i,2]))[10:12,1]
      desol  <- data.matrix(extractCTDC (protein_sequence[i,2]))[19:21,1]

    #add new data
    }else{
      protein_name  <- rbind(protein_name, protein_sequence[i,1],row.names = NULL)
      hydro  <- cbind(hydro, data.matrix(extractCTDC (protein_sequence[i,2]))[1:3,1],row.names = NULL)
      vdm  <- cbind(vdm, data.matrix(extractCTDC (protein_sequence[i,2]))[4:6,1],row.names = NULL)
      pol  <- cbind(pol, data.matrix(extractCTDC (protein_sequence[i,2]))[7:9,1],row.names = NULL)
      polabil  <- cbind(polabil, data.matrix(extractCTDC (protein_sequence[i,2]))[10:12,1],row.names = NULL)
      desol  <- cbind(desol, data.matrix(extractCTDC (protein_sequence[i,2]))[19:21,1],row.names = NULL)
    }

  }
  attr_name  <- switch(type,"hydrophobicity","VWF Volume","polarity", "polarizability","desolvation")
  #different attribute data selected according to type
  len_type  <- switch(type, hydro, vdm, pol, polabil, desol)
  #adding category name
  category_name  <- switch(type,
                    c("Polar","Neutral","Hydrophobic"),
                    c("van der Waals 0-2.78 ",
                      "van der Waals 2.95-4.0",
                      "van der Waals 4.03-8.08"),
                    c("polarity 4.9-6.2 ",
                      "polarity 8.0-9.2",
                      "polarity 10.4-13.0"),
                    c("Polarizability 0-1.08 ",
                      "Polarizability 0.128-0.186",
                      "Polarizability 0.219-0.409"),
                    c("Buried ","Exposed","Intermediate"))
  #3 sub categories for each category
  protein_data=rep(protein_name, each=3)

  len=c(len_type)

  #construct dataframe for plotting
  dataf <- data.frame(protein_data=rep(c(protein_name), each=3),
                      category=rep(category_name, protein_num),
                      len=hydro, row.names =NULL,stringsAsFactors=FALSE)

  #circular plot option
  #Some code in circular plot on formatting (clean grid and adjust size)
  #are adapted from the R graph gallery
  #https://www.r-graph-gallery.com/297-circular-barplot-with-groups.html
  if(circular_plot){
    p  <- ggplot(data = dataf,
               aes(x = protein_data, y = len, fill = category)) +
      geom_bar(stat = "identity") + ylim(-1, 1) + theme_minimal() +
      theme(
            plot.title = element_text(hjust = 0.5),
            axis.title = element_blank(),
            panel.grid = element_blank(), ) +
      coord_polar(start = 0) +
      ggtitle(paste("Protein Sequence composition of ",attr_name,sep = ""))
  # normal plot option
  }else{
    p <- ggplot(data=dataf,
               aes(x=protein_data, y=len, fill=category)) +
      geom_bar(stat="identity") +
      ggtitle(paste("Protein Sequence composition of ",attr_name,sep = ""))

  }

  return(p)


}
#Reference
#The R Graph Gallery -Circular barplot with groups. (2018).
#Retrieved December 1, 2019, from https://www.r-graph-gallery.com/297-circular-barplot-with-groups.html



inst/shiny-scripts/Rppsc.R
library(shiny)
library(Rppsc)

ui <- fluidPage(

  titlePanel("Rppsc: R Package for Plotting Protein Composition"),
  #File input box
  fileInput(inputId = "file1",
            label = "Select your sequence.fasta file",
            multiple = FALSE),
  #instruction text
  textOutput("txt"),
  #5 button for 5 different kind of composition map
  actionButton("hydrophobicity", "hydrophobicity"),
  actionButton("Van_Der_Waals", "Van Der Waals"),
  actionButton("polarity", "polarity"),
  actionButton("polarizability", "polarizability"),
  actionButton("desolvation", "desolvation"),
  hr(),
  plotOutput("composition_plot")
)

server <- function(input, output){
  #initialize a default vector for the function parameters
  v <- reactiveValues(
    filepath = "proSeq",
    type = NULL,
    circular_plot = TRUE
    )
  #change type when user hit different button
  observeEvent(input$hydrophobicity, {
    v$type <- 1
  })
  observeEvent(input$Van_Der_Waals, {
    v$type <- 2
  })
  observeEvent(input$polarity, {
    v$type <- 3
  })
  observeEvent(input$polarizability, {
    v$type <- 4
  })
  observeEvent(input$desolvation, {
    v$type <- 5
  })
  #generate the plot
  output$composition_plot <- renderPlot({
    if (is.null(v$type)) return()
    plotCG(file = v$filepath, type = v$type, circular_plot = v$circular_plot)

  })

  output$txt <- renderText({
    paste("Click the following button to see the corresponding protein composition")
  })
}

shinyApp(ui, server)



R/plotAPCG.R
#' Plot All Protein Composition Graph
#'
#' A function that plots a comprehensive composition mapping given a protein list.
#' The composition descriptors includes hydrophobicity, van der Waals volume,
#' polarity, polarizability and desolvation. A hierarchical row and column clustering
#' will also be displayed on the side to showing the similarity analysis
#'
#' @param file input csv file containing the list of pdb id and sequence
#'
#' @return
#' This function returns a matrix of composition descriptor infomation as well as
#' a full composition graph given the proteins.
#'
#' @examples
#' composition_plot <- plotAPCG(file = "proSeq")
#' composition_plot
#'
#' @references
#'Xiao, N., Cao, D.-S., Zhu, M.-F., & Xu, Q.-S. (2015). protr/ProtrWeb: R package and web server for
#'generating various numerical representation schemes of protein sequences. Bioinformatics, 31(11), 1857–1859.
#'https://doi.org/10.1093/bioinformatics/btv042
#'
#'
#' @export
#' @import protr
#' @import gplots
#' @import utils
plotAPCG <- function(file = "proSeq") {
  library(protr)
  library(gplots)
  library(utils)

  #Validate file path
  if(!(file_test("-f",file)) & (file != "proSeq")){
    stop("Invalid file path.")
  }

  #load the default data
  if (file == "proSeq"){
    protein_sequence = data.frame(lapply(proSeq, as.character), stringsAsFactors=FALSE)
  #load data from validated file path
  }else{
    raw_sequence = read.csv(filepath,header = TRUE,sep = ",")
    protein_sequence = data.frame(lapply(raw_sequence, as.character), stringsAsFactors=FALSE)
  }
  #five attribute for analysis
  colname = c("hydrophobicity","VWF Volume","polarity", "polarizability","desolvation")
  #the full info of five attributes
  five_attr_info = NULL
  #loop over all proteins to calculate the average of each attributes
  for(i in seq_along(protein_sequence[,1]) ) {
    #pdbid
    pdb = protein_sequence[i,1]
    #sequence
    pseq = protein_sequence[i,2]
    #protein sequence composition data generated by extractDTDC()
    composition_data = extractCTDC (pseq)
    # Compute average of 5 categories: hydrophobiciy, van der Waals volume, polarity
    #polarizability,desolvation.
    # The classification and calculation is based on the Xiao's article on the three-group classification of the 20 amino acids
    # by each attribute.
    hydro = composition_data[[3]]
    vdw = (composition_data[[4]]*2.78+composition_data[[5]]*4+composition_data[[6]]*8.08)/3
    pol= (composition_data[[7]]*6.2+composition_data[[8]]*9.2+composition_data[[9]]*13.0)/3
    polarizability = (composition_data[[10]]*1.08+composition_data[[11]]*0.186+composition_data[[12]]*0.409)/3
    desol = composition_data[[19]]

    #initiation
    if (is.null(five_attr_info)){
      five_attr_info = matrix(c(hydro,vdw, pol, polarizability, desol), nrow = 1,ncol= 5, dimnames = list(pdb, colname))
    #add new data
    }else{
      subm = matrix(c(hydro,vdw, pol, polarizability, desol), nrow = 1,ncol= 5, dimnames = list(pdb, colname))
      five_attr_info = rbind(five_attr_info,subm)
    }

  }
  #construct dataframe
  dataf = as.data.frame(five_attr_info)
  #Heatmap of the result
  mapping_result = heatmap.2(scale(five_attr_info), cexCol = 1.2, margins = c(7.5, 4))

  return(mapping_result)

}

#Reference
#Xiao, N., Cao, D.-S., Zhu, M.-F., & Xu, Q.-S. (2015). protr/ProtrWeb: R package and web server for
#generating various numerical representation schemes of protein sequences. Bioinformatics, 31(11), 1857–1859.
#https://doi.org/10.1093/bioinformatics/btv042
